aicsimageio/writers/writer.py

```diff
@@ -1,49 +1,78 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
 from abc import ABC, abstractmethod
+from typing import Any, Optional, Union
 
-from .. import constants, types
+import dask.array as da
+import numpy as np
 
+from ..constants import Dimensions
 
-class Writer(ABC):
-    def __init__(self, file_path: types.PathLike):
-        """
-        Write STCZYX data arrays to file, with accompanying metadata
-        Will overwrite existing files of same name.
+###############################################################################
 
-        Parameters
-        ----------
-        file_path: types.PathLike
-            Path to image output location
 
-        Examples
-        --------
-        Construct and use as object
+class Writer(ABC):
+    """
+    Write a data array to file, with accompanying metadata
+    This will overwrite existing files of same name.
 
-        >>> image = numpy.ndarray([1, 10, 3, 1024, 2048])
-        ... writer = DerivedWriter("file.ome.tif")
-        ... writer.save(image)
-        ... writer.close()
+    Parameters
+    ----------
+    filepath: types.PathLike
+        Path to image output location
 
-        Construct with a context manager
+    Examples
+    --------
+    Construct and use as object
+    >>> image = numpy.ndarray([1, 10, 3, 1024, 2048])
+    ... writer = DerivedWriter("file.ome.tif")
+    ... writer.save(image)
+    ... writer.close()
 
-        >>> image2 = numpy.ndarray([5, 486, 210])
-        ... with DerivedWriter("file2.ome.tif") as writer2:
-        ...     writer2.set_metadata(myMetaData)
-        ...     writer2.save(image2)
-        """
-        self.file_path = file_path
+    Construct with a context manager
+    >>> image2 = numpy.ndarray([5, 486, 210])
+    ... with DerivedWriter("file2.ome.tif") as writer2:
+    ...     writer2.save(image2, metadata=some_metadata_object)
+    """
+    filepath = None
+    _buffer = None
```

Comment:
This rather than `__init__` to stay consistent with `Reader` and technically you can't initialize a base class.

Reply:
NIT: you have to declare __init__ an @abstractmethod for it to be uninstantiable in python.
If this is what you want.

Reply:
```
In [1]: from aicsimageio.writers.writer import Writer                                                                                                                                         

In [2]: w = Writer()                                                                                                                                                                          
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-2-fec5b6797a9a> in <module>
----> 1 w = Writer()

TypeError: Can't instantiate abstract class Writer with abstract methods close, open, save
```

Nope can't instantiate a base class at all because _any_ of the methods are abstract.

Reply:
ok that's weird, before making the comment I tested it on 3.7 and created an ABC and then was able to instantiate it. Maybe I made a mistake but if it does throw an exception that's definitely preferable.